components/script/fetch.rs

```diff
@@ -266,6 +273,201 @@ pub(crate) fn Fetch(
+    // TODO
+    // Step 2. Set request’s service-workers mode to "none".
+    // TODO
+    // Step 3. Set request’s keepalive to true.
+    // TODO
+    // Step 4. Let deferredRecord be a new deferred fetch record whose request is request, and whose notify invoked is onActivatedWithoutTermination.
+    let deferred_record = Arc::new(Mutex::new(DeferredFetchRecord {
+        request,
+        global: trusted_global.clone(),
+        invoke_state: Cell::new(DeferredFetchRecordInvokeState::Pending),
+        activated: Cell::new(false),
+    }));
+    // Step 5. Append deferredRecord to request’s client’s fetch group’s deferred fetch records.
+    // TODO
+    // Step 6. If activateAfter is non-null, then run the following steps in parallel:
+    let now = Instant::now();
+    let deferred_record_clone = deferred_record.clone();
+    global.task_manager().deferred_fetch_task_source().queue(
+        task!(process_deferred_record: move || {
+            // Step 6.1. The user agent should wait until any of the following conditions is met:
+            let elapsed = now.elapsed().as_millis() as u64;
+            let activate_after = *activate_after as u64;
+            if elapsed < activate_after {
+                let sleep_duration = activate_after - elapsed;
+                thread::sleep(Duration::from_millis(sleep_duration));
+            }
+            // Step 6.2. Process deferredRecord.
+            deferred_record_clone.lock().unwrap().process();
+
+            // Last step of https://fetch.spec.whatwg.org/#process-a-deferred-fetch
+            //
+            // Step 4. Queue a global task on the deferred fetch task source with
+            // deferredRecord’s request’s client’s global object to run deferredRecord’s notify invoked.
+            trusted_global
+                .root()
+                .task_manager()
+                .deferred_fetch_task_source()
+                .queue(task!(notify_deferred_record: move || {
+                    deferred_record_clone.lock().unwrap().activate();
+                }));
+        }),
+    );
+    // Step 7. Return deferredRecord.
+    deferred_record
+}
+
+/// <https://fetch.spec.whatwg.org/#dom-window-fetchlater>
+#[allow(non_snake_case, unsafe_code)]
+pub(crate) fn FetchLater(
+    window: &Window,
+    input: RequestInfo,
+    init: RootedTraceableBox<DeferredRequestInit>,
+    can_gc: CanGc,
+) -> Fallible<DomRoot<FetchLaterResult>> {
+    let global_scope = window.upcast();
+    // Step 1. Let requestObject be the result of invoking the initial value
+    // of Request as constructor with input and init as arguments.
+    let request_object = Request::constructor(global_scope, None, can_gc, input, &init.parent)?;
+    // Step 2. If requestObject’s signal is aborted, then throw signal’s abort reason.
+    let signal = request_object.Signal();
+    if signal.aborted() {
+        // TODO: Throw abort reason here without also returning an `Err`
```

Comment:
I tried to update `Error` to take a `HandleValue`, so that I don't have to return twice. Reason being that after I call `report_pending_exception` I would still need to return something. I can't return `Ok`, since that requires creating a `FetchLaterResult`. And I can't return `Err` since that would cause two exceptions to be thrown.

However, `HandleValue` has a lifetime, so can't add that to `Error` either. For now we always throw an AbortError, but that's wrong and we should throw the signal instead.

Reply:
If I understand your problem correctly then this is precisely what [`Error::JSFailed`](https://doc.servo.org/script/dom/bindings/error/enum.Error.html#variant.JSFailed) is for, you can return that.

Reply:
Yes that works, thanks!